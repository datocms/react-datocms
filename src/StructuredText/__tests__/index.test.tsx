import { renderMarkRule } from "datocms-structured-text-generic-html-renderer";
import { isHeading } from "datocms-structured-text-utils";
import { mount, render, shallow } from "enzyme";
import * as React from "react";
import {
  RenderError,
  StructuredText,
  type StructuredTextDocument,
  type StructuredTextGraphQlResponse,
  renderNodeRule,
} from "../index.js";

describe("StructuredText", () => {
  describe("with no value", () => {
    it("renders null", () => {
      const wrapper = mount(<StructuredText data={null} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("simple dast /2", () => {
    const structuredText: StructuredTextDocument = {
      schema: "dast",
      document: {
        type: "root",
        children: [
          {
            type: "heading",
            level: 1,
            children: [
              {
                type: "span",
                value: "This\nis a ",
              },
              {
                type: "span",
                marks: ["strong"],
                value: "title",
              },
            ],
          },
        ],
      },
    };

    describe("with default rules", () => {
      it("renders the document", () => {
        const wrapper = mount(<StructuredText data={structuredText} />);
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe("with custom mark rules", () => {
      it("renders the document", () => {
        const wrapper = mount(
          <StructuredText
            data={structuredText}
            customMarkRules={[
              renderMarkRule("strong", ({ children, key }) => (
                <b key={key}>{children}</b>
              )),
            ]}
          />
        );
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe("simple dast with no links/blocks", () => {
    const structuredText: StructuredTextGraphQlResponse = {
      value: {
        schema: "dast",
        document: {
          type: "root",
          children: [
            {
              type: "heading",
              level: 1,
              children: [
                {
                  type: "span",
                  value: "This\nis a ",
                },
                {
                  type: "span",
                  marks: ["strong"],
                  value: "title",
                },
              ],
            },
          ],
        },
      },
    };

    describe("with default rules", () => {
      it("renders the document", () => {
        const wrapper = mount(<StructuredText data={structuredText} />);
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe("with custom rules", () => {
      it("renders the document", () => {
        const wrapper = mount(
          <StructuredText
            data={structuredText}
            renderText={(text, key) => {
              return (
                <React.Fragment key={key}>
                  {text.replace(/This/, "That")}
                </React.Fragment>
              );
            }}
            customNodeRules={[
              renderNodeRule(
                isHeading,
                ({ adapter: { renderNode }, node, children, key }) => {
                  return renderNode(`h${node.level + 1}`, { key }, children);
                }
              ),
            ]}
          />
        );
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe("with links/blocks", () => {
    type QuoteRecord = {
      id: string;
      __typename: "QuoteRecord";
      quote: string;
      author: string;
    };

    type MentionRecord = {
      id: string;
      __typename: "MentionRecord";
      name: string;
    };

    type DocPageRecord = {
      id: string;
      __typename: "DocPageRecord";
      slug: string;
      title: string;
    };

    const structuredText: StructuredTextGraphQlResponse<
      QuoteRecord | MentionRecord,
      DocPageRecord
    > = {
      value: {
        schema: "dast",
        document: {
          type: "root",
          children: [
            {
              type: "heading",
              level: 1,
              children: [
                {
                  type: "span",
                  value: "This is a",
                },
                {
                  type: "span",
                  marks: ["highlight"],
                  value: "title",
                },
                {
                  type: "inlineItem",
                  item: "123",
                },
                {
                  type: "itemLink",
                  item: "123",
                  meta: [{ id: "target", value: "_blank" }],
                  children: [{ type: "span", value: "here!" }],
                },
                {
                  type: "inlineBlock",
                  item: "789",
                },
              ],
            },
            {
              type: "block",
              item: "456",
            },
          ],
        },
      },
      blocks: [
        {
          id: "456",
          __typename: "QuoteRecord",
          quote: "Foo bar.",
          author: "Mark Smith",
        },
        {
          id: "789",
          __typename: "MentionRecord",
          name: "Jane Doe",
        },
      ],
      links: [
        {
          id: "123",
          __typename: "DocPageRecord",
          title: "How to code",
          slug: "how-to-code",
        },
      ],
    };

    describe("with default rules", () => {
      it("renders the document", () => {
        const wrapper = mount(
          <StructuredText
            data={structuredText}
            renderInlineRecord={({ record }) => {
              switch (record.__typename) {
                case "DocPageRecord":
                  return <a href={`/docs/${record.slug}`}>{record.title}</a>;
                default:
                  return null;
              }
            }}
            renderLinkToRecord={({ record, children, transformedMeta }) => {
              switch (record.__typename) {
                case "DocPageRecord":
                  return (
                    <a {...transformedMeta} href={`/docs/${record.slug}`}>
                      {children}
                    </a>
                  );
                default:
                  return null;
              }
            }}
            renderBlock={({ record }) => {
              switch (record.__typename) {
                case "QuoteRecord":
                  return (
                    <figure>
                      <blockquote>{record.quote}</blockquote>
                      <figcaption>{record.author}</figcaption>
                    </figure>
                  );
                default:
                  return null;
              }
            }}
            renderInlineBlock={({ record }) => {
              switch (record.__typename) {
                case "MentionRecord":
                  return <em>{record.name}</em>;
                default:
                  return null;
              }
            }}
          />
        );
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe("with missing renderInlineRecord prop", () => {
      it("raises an error", () => {
        expect(() => {
          shallow(<StructuredText data={structuredText} />);
        }).toThrow(RenderError);
      });
    });

    describe("with missing record", () => {
      it("raises an error", () => {
        expect(() => {
          shallow(
            <StructuredText
              data={{ ...structuredText, links: [] }}
              renderInlineRecord={() => {
                return null;
              }}
            />
          );
        }).toThrow(RenderError);
      });
    });
  });
});
