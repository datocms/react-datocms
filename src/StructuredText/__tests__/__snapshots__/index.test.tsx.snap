// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StructuredText simple dast /2 with custom mark rules renders the document 1`] = `
<StructuredText
  customMarkRules={
    [
      {
        "appliable": [Function],
        "apply": [Function],
      },
    ]
  }
  data={
    {
      "document": {
        "children": [
          {
            "children": [
              {
                "type": "span",
                "value": "This
is a ",
              },
              {
                "marks": [
                  "strong",
                ],
                "type": "span",
                "value": "title",
              },
            ],
            "level": 1,
            "type": "heading",
          },
        ],
        "type": "root",
      },
      "schema": "dast",
    }
  }
>
  <h1
    key="t-0"
  >
    This
    <br
      key="t-0-br-0"
    />
    is a 
    <b
      key="t-1"
    >
      title
    </b>
  </h1>
</StructuredText>
`;

exports[`StructuredText simple dast /2 with default rules renders the document 1`] = `
<StructuredText
  data={
    {
      "document": {
        "children": [
          {
            "children": [
              {
                "type": "span",
                "value": "This
is a ",
              },
              {
                "marks": [
                  "strong",
                ],
                "type": "span",
                "value": "title",
              },
            ],
            "level": 1,
            "type": "heading",
          },
        ],
        "type": "root",
      },
      "schema": "dast",
    }
  }
>
  <h1
    key="t-0"
  >
    This
    <br
      key="t-0-br-0"
    />
    is a 
    <strong
      key="t-1"
    >
      title
    </strong>
  </h1>
</StructuredText>
`;

exports[`StructuredText simple dast with no links/blocks with custom rules renders the document 1`] = `
<StructuredText
  customNodeRules={
    [
      {
        "appliable": [Function],
        "apply": [Function],
      },
    ]
  }
  data={
    {
      "value": {
        "document": {
          "children": [
            {
              "children": [
                {
                  "type": "span",
                  "value": "This
is a ",
                },
                {
                  "marks": [
                    "strong",
                  ],
                  "type": "span",
                  "value": "title",
                },
              ],
              "level": 1,
              "type": "heading",
            },
          ],
          "type": "root",
        },
        "schema": "dast",
      },
    }
  }
  renderText={[Function]}
>
  <h2
    key="t-0"
  >
    That
    <br
      key="t-0-br-0"
    />
    is a 
    <strong
      key="t-1"
    >
      title
    </strong>
  </h2>
</StructuredText>
`;

exports[`StructuredText simple dast with no links/blocks with default rules renders the document 1`] = `
<StructuredText
  data={
    {
      "value": {
        "document": {
          "children": [
            {
              "children": [
                {
                  "type": "span",
                  "value": "This
is a ",
                },
                {
                  "marks": [
                    "strong",
                  ],
                  "type": "span",
                  "value": "title",
                },
              ],
              "level": 1,
              "type": "heading",
            },
          ],
          "type": "root",
        },
        "schema": "dast",
      },
    }
  }
>
  <h1
    key="t-0"
  >
    This
    <br
      key="t-0-br-0"
    />
    is a 
    <strong
      key="t-1"
    >
      title
    </strong>
  </h1>
</StructuredText>
`;

exports[`StructuredText with links/blocks with default rules renders the document 1`] = `
<StructuredText
  data={
    {
      "blocks": [
        {
          "__typename": "QuoteRecord",
          "author": "Mark Smith",
          "id": "456",
          "quote": "Foo bar.",
        },
      ],
      "inlineBlocks": [
        {
          "__typename": "MentionRecord",
          "id": "789",
          "name": "Jane Doe",
        },
      ],
      "links": [
        {
          "__typename": "DocPageRecord",
          "id": "123",
          "slug": "how-to-code",
          "title": "How to code",
        },
      ],
      "value": {
        "document": {
          "children": [
            {
              "children": [
                {
                  "type": "span",
                  "value": "This is a",
                },
                {
                  "marks": [
                    "highlight",
                  ],
                  "type": "span",
                  "value": "title",
                },
                {
                  "item": "123",
                  "type": "inlineItem",
                },
                {
                  "children": [
                    {
                      "type": "span",
                      "value": "here!",
                    },
                  ],
                  "item": "123",
                  "meta": [
                    {
                      "id": "target",
                      "value": "_blank",
                    },
                  ],
                  "type": "itemLink",
                },
                {
                  "item": "789",
                  "type": "inlineBlock",
                },
              ],
              "level": 1,
              "type": "heading",
            },
            {
              "item": "456",
              "type": "block",
            },
          ],
          "type": "root",
        },
        "schema": "dast",
      },
    }
  }
  renderBlock={[Function]}
  renderInlineBlock={[Function]}
  renderInlineRecord={[Function]}
  renderLinkToRecord={[Function]}
>
  <h1
    key="t-0"
  >
    This is a
    <mark
      key="t-1"
    >
      title
    </mark>
    <a
      href="/docs/how-to-code"
      key="t-2"
    >
      How to code
    </a>
    <a
      href="/docs/how-to-code"
      key="t-3"
      target="_blank"
    >
      here!
    </a>
    <em
      key="t-4"
    >
      Jane Doe
    </em>
  </h1>
  <figure
    key="t-1"
  >
    <blockquote>
      Foo bar.
    </blockquote>
    <figcaption>
      Mark Smith
    </figcaption>
  </figure>
</StructuredText>
`;

exports[`StructuredText with no value renders null 1`] = `
<StructuredText
  data={null}
/>
`;
